# Day 1: Calorie Counting

Santa's reindeer typically eat regular reindeer food, but they need a lot of magical energy to deliver presents on Christmas. For that, their favorite snack is a special type of star fruit that only grows deep in the jungle. The Elves have brought you on their annual expedition to the grove where the fruit grows.

To supply enough magical energy, the expedition needs to retrieve a minimum of fifty stars by December 25th. Although the Elves assure you that the grove has plenty of fruit, you decide to grab any fruit you see along the way, just in case.

Collect stars by solving puzzles. Two puzzles will be made available on each day in the Advent calendar; the second puzzle is unlocked when you complete the first. Each puzzle grants one star. Good luck!

The jungle must be too overgrown and difficult to navigate in vehicles or access from the air; the Elves' expedition traditionally goes on foot. As your boats approach land, the Elves begin taking inventory of their supplies. One important consideration is food - in particular, the number of Calories each Elf is carrying (your puzzle input).

The Elves take turns writing down the number of Calories contained by the various meals, snacks, rations, etc. that they've brought with them, one item per line. Each Elf separates their own inventory from the previous Elf's inventory (if any) by a blank line.

For example, suppose the Elves finish writing their items' Calories and end up with the following list:

```
1000
2000
3000

4000

5000
6000

7000
8000
9000

10000
```

This list represents the Calories of the food carried by five Elves:

- The first Elf is carrying food with 1000, 2000, and 3000 Calories, a total of 6000 Calories.
- The second Elf is carrying one food item with 4000 Calories.
- The third Elf is carrying food with 5000 and 6000 Calories, a total of 11000 Calories.
- The fourth Elf is carrying food with 7000, 8000, and 9000 Calories, a total of 24000 Calories.
- The fifth Elf is carrying one food item with 10000 Calories.

In case the Elves get hungry and need extra snacks, they need to know which Elf to ask: they'd like to know how many Calories are being carried by the Elf carrying the most Calories. In the example above, this is 24000 (carried by the fourth Elf).

Find the Elf carrying the most Calories. How many total Calories is that Elf carrying?

In [96]:
#@title Import packages
import numpy as np
import pandas as pd

In [97]:
#@title import the data
dfx = None
input_dataset = []
with open('input.txt', 'r') as f:
    array_entry = []
    for line in f.readlines():
        if line == '\n':
            input_dataset.append(array_entry)
            array_entry = []
        else:    
            array_entry.append(float(line))

dfx = pd.DataFrame(input_dataset)

In [98]:
#@title Display imported data
dfx

0        1        2        3       4       5       6       7   \
0    7914.0   5032.0  11424.0   2567.0  4123.0  3567.0  7346.0     NaN   
1    1334.0   2173.0   5437.0   1104.0  1872.0  1148.0  6547.0  3149.0   
2    4108.0  14444.0  25596.0      NaN     NaN     NaN     NaN     NaN   
3    7237.0   4252.0   1020.0   8333.0  3538.0  5089.0  6328.0  4581.0   
4    5799.0   1585.0   1137.0   1985.0  6489.0  5309.0  6249.0  4594.0   
..      ...      ...      ...      ...     ...     ...     ...     ...   
247  3334.0   8211.0   5328.0   4224.0  5626.0  7635.0  7615.0  9511.0   
248  8075.0   5712.0   7226.0   5213.0  7236.0  7907.0  1417.0  6890.0   
249  1189.0   2124.0   5640.0   4677.0  1487.0  3608.0  2828.0  2911.0   
250  7302.0   1351.0   1299.0   3498.0  2752.0  2502.0  7397.0  7139.0   
251  2453.0   2774.0  12613.0  15345.0  3686.0     NaN     NaN     NaN   

         8       9       10      11      12      13      14  
0       NaN     NaN     NaN     NaN     NaN     NaN     NaN  
1    5923.0  5705.0  4036.0  5348.0  1100.0     NaN     NaN  
2       NaN     NaN     NaN     NaN     NaN     NaN     NaN  
3    6503.0  8741.0     NaN     NaN     NaN     NaN     NaN  
4    1885.0  6100.0  4233.0  1517.0  6346.0     NaN     NaN  
..      ...     ...     ...     ...     ...     ...     ...  
247     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
248  2571.0  1881.0  6929.0     NaN     NaN     NaN     NaN  
249  2102.0  2169.0  6087.0  4343.0  5689.0  5315.0  1946.0  
250  1571.0  1882.0  4967.0  7024.0     NaN     NaN     NaN  
251     NaN     NaN     NaN     NaN     NaN     NaN     NaN  

[252 rows x 15 columns]

In [99]:
#@title Clean the data

# Add Column Names
column_names = [f'Item {i+1}' for i in range(len(dfx.columns))]
dfx.columns = column_names

# Replace index from 0- to 1-indexed
dfx.index = range(1, len(dfx) + 1)

# Convert NaN to 0
dfx = dfx.fillna(0)

In [103]:
#@title Add new 'Sum' column
dfx["Sum"] = dfx.sum(axis = 1) # axis to sum rows not default of columns

In [108]:
#@title Get the answer
dfx.sort_values('Sum', ascending=False).head(1)["Sum"]

127    64929.0
Name: Sum, dtype: float64

Your puzzle answer was 64929.

The first half of this puzzle is complete! It provides one gold star: *

## Part Two

By the time you calculate the answer to the Elves' question, they've already realized that the Elf carrying the most Calories of food might eventually run out of snacks.

To avoid this unacceptable situation, the Elves would instead like to know the total Calories carried by the top three Elves carrying the most Calories. That way, even if one of those Elves runs out of snacks, they still have two backups.

In the example above, the top three Elves are the fourth Elf (with 24000 Calories), then the third Elf (with 11000 Calories), then the fifth Elf (with 10000 Calories). The sum of the Calories carried by these three elves is 45000.

Find the top three Elves carrying the most Calories. How many Calories are those Elves carrying in total?

In [109]:
#@title Get the answer to part 2
dfx.sort_values('Sum', ascending=False).head(3)["Sum"].sum()

193697.0

That's the right answer! You are one gold star closer to collecting enough star fruit.

**You have completed Day 1!**